Tests/kaas/k8s-version-recency/k8s-version-recency-check.py

```diff
@@ -0,0 +1,465 @@
+
+    def __lt__(self, other):
+        if not isinstance(other, K8sVersionInfo):
+            return False
+        patchcomp = self.minor == other.minor and self.patch < other.patch
+        return self.major < other.major or (self.major == other.major and (self.minor < other.minor or patchcomp))
+
+    def __le__(self, other):
+        if not isinstance(other, K8sVersionInfo):
+            return False
+        patchcomp = self.minor == other.minor and self.patch <= other.patch
+        return self.major < other.major or (self.major == other.major and (self.minor < other.minor or patchcomp))
+
+    @classmethod
+    def extract_version(cls, string, separator="."):
+        components = string.strip().split(separator)
+        return cls(int(components[0]), int(components[1]), int(components[2]))
+
+    def check_for_version(self, major=None, minor=None, patch=None):
+        return (major is None or self.major == major) and \
+               (minor is None or self.minor == minor) and \
+               (patch is None or self.patch == patch)
+
+    def __str__(self):
+        return f"{self.major}.{self.minor}.{self.patch}"
+
+
+class CVEVersionInfo:
+    def __init__(self, lower_version, upper_version, less_than=False, equal=False):
+        self.lower_version = lower_version
+        self.upper_version = upper_version
+
+        self.less_than = less_than
+        self.equal = equal
+
+    def __eq__(self, other):
+        if not isinstance(other, CVEVersionInfo):
+            return False
+        return self.lower_version == other.lower_version and self.upper_version == other.upper_version and \
+            self.less_than == other.less_than and self.equal == self.equal
+
+    def is_version_affected(self, version_info):
+        if self.upper_version:
+            if self.lower_version:
+                gt = self.lower_version <= version_info
+            else:
+                gt = True
+            if self.less_than:
+                if self.equal:
+                    return gt and self.upper_version >= version_info
+                return gt and self.upper_version > version_info
+        else:
+            return self.lower_version == version_info
+
+
+def diff_months(date1, date2):
+    return abs((date1.year - date2.year) * 12 + date1.month - date2.month)
+
+
+def diff_weeks(date1, date2):
+    day1 = (date1 - datetime.timedelta(days=date1.weekday()))
+    day2 = (date2 - datetime.timedelta(days=date2.weekday()))
+    diff = day2 - day1
+    return abs((diff.days / 7) + math.ceil(diff.seconds / 86400))
+
+
+def diff_days(date1, date2):
+    delta = date1 - date2
+    return abs(delta.days)
+
+
+async def request_cve_data(session: aiohttp.ClientSession, cveid: str) -> dict:
+    """Request for a single CVE data item."""
+    resp = await session.request('GET', f"https://cveawg.mitre.org/api/cve/{cveid}",
+                                 headers={"Accept": "application/json"})
+    return await resp.json()
```

Comment:
Updated to your suggestion.

Reply:
@fdobrovolny  Sorry for interjecting here, but out of curiosity, what is the difference?

Reply:
I think it does some logic to close the connection as the `with` operator calls "`open`" and "`close`" on the object.

I was just going out of the documentation and how it is intended to be used, never questioning it. If you want to take a look at the source code, I would also be interested.

https://docs.aiohttp.org/en/stable/client_quickstart.html#make-a-request

> Now, let’s try to get a web-page. For example let’s query http://httpbin.org/get:
> 
> ```
> async def main():
>     async with aiohttp.ClientSession() as session:
>         async with session.get('http://httpbin.org/get') as resp:
>             print(resp.status)
>             print(await resp.text())
> 
> asyncio.run(main())
> ```

> To make several requests to the same site more simple, the parameter base_url of [ClientSession](https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession) constructor can be used. For example to request different endpoints of http://httpbin.org can be used the following code:
> ```
> async with aiohttp.ClientSession('http://httpbin.org') as session:
>     async with session.get('/get'):
>         pass
>     async with session.post('/post', data=b'data'):
>         pass
>     async with session.put('/put', data=b'data'):
>         pass
> ```

> A session context manager usage is not mandatory but await session.close() method should be called in this case, e.g.:
> 
> ```
> session = aiohttp.ClientSession()
> async with session.get('...'):
>     # ...
> await session.close()
> ```

Reply:
Interesting that the quickstart docs should recommend that, because the [reference](https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientResponse.release) also states

> It is not required to call release on the response object. When the client fully receives the payload, the underlying connection automatically returns back to pool. If the payload is not fully read, the connection is closed

So it seems to me that the `async with` is kind of redundant for us, because we do consume the payload (by calling `json`).

At any rate, it has nothing to do with being async, and more with returning the resources (the connection in this case) to the pool. It may be more economical with the context manager (though, as mentioned, not in our case), but it's not "more async".

Reply:
Indeed, it is interesting. I find some philosophical value in it as well, as it promotes a consistent approach that always works, aligning with The Zen of Python principles.

I meant "properly async" to suggest that either the python `requests` library should be used correctly or the `aiohttp` library should be adopted fully. At the time, it appeared to be a partial transition to async(`aiohttp`) based on the quickstart and my experience, hence the term "...properly async it should look more like this." However, since it was functioning, it was meant as a suggestion rather than a requirement.

Reply:
Can this be resolved?